astropy/utils/decorators.py

```diff
@@ -751,3 +752,229 @@ def _get_function_args(func, exclude_args=()):
+            Perform addition.
+            Parameters
+            ----------
+            num1, num2 : Numbers
+            Returns
+            -------
+            result: Number
+                result of num1 + num2
+        >>> help(subtract) # doctest: +SKIP
+        Help on function subtract in module __main__:
+        <BLANKLINE>
+        subtract(num1, num2)
+            Perform subtraction.
+            Parameters
+            ----------
+            num1, num2 : Numbers
+            Returns
+            -------
+            result: Number
+                result of num1 - num2
+            Notes: This one has additional notes.
+
+    These methods can be combined an even taking the docstring from another
+    object is possible as docstring attribute. You just have to specify the
+    object::
+
+        >>> @replace_and_format_docstring(add)
+        ... def another_add(num1, num2):
+        ...     return num1 + num2
+        ...
+        >>> help(another_add) # doctest: +SKIP
+        Help on function another_add in module __main__:
+        <BLANKLINE>
+        another_add(num1, num2)
+            Perform addition.
+            Parameters
+            ----------
+            num1, num2 : Numbers
+            Returns
+            -------
+            result: Number
+                result of num1 + num2
+
+    But be aware that this decorator *only* formats the given docstring not
+    the docstring that is formulated inside the function itself::
+
+        >>> @replace_and_format_docstring(doc, 'addition', op='+')
+        ... def yet_another_add(num1, num2):
+        ...    '''This one is good for {0}.'''
+        ...    return num1 + num2
+        ...
+        >>> help(yet_another_add) # doctest: +SKIP
+        Help on function yet_another_add in module __main__:
+        <BLANKLINE>
+        yet_another_add(num1, num2)
+            Perform addition.
+            Parameters
+            ----------
+            num1, num2 : Numbers
+            Returns
+            -------
+            result: Number
+                result of num1 + num2
+            This one is good for {0}.
+
+    To work around it you could specify the docstring to be ``'self'``
+    but then the ``args`` and ``kwargs`` don't get formatted::
+
+        >>> @replace_and_format_docstring('self', 'addition')
+        ... def last_add_i_swear(num1, num2):
+        ...    '''This one is good for {0}.'''
+        ...    return num1 + num2
+        ...
+        >>> help(last_add_i_swear) # doctest: +SKIP
+        Help on function last_add_i_swear in module __main__:
+        <BLANKLINE>
+        last_add_i_swear(num1, num2)
+            This one is good for addition.
+
+    Using it with ``'self'`` as docstring allows to use the decorator twice on
+    an object to first parse the new docstring and then to parse the original
+    docstring. But since this decorator performs string operations using it
+    too often might affect the performance.
```

Comment:
Why might performance be an issue? This is done only when constructing the function/method, correct? I.e., it only affects setup, not runtime performance? If so, I'd remove this last sentence.